modules/adbookpspBidAdapter.js

```diff
@@ -0,0 +1,796 @@
+  bidderRequestBody,
+  bidResponseCurrency,
+  referrer,
+  targetingMap
+) {
+  return incomingBids
+    .map(
+      impToPrebidBid(
+        bidderRequestBody,
+        bidResponseCurrency,
+        referrer,
+        targetingMap
+      )
+    )
+    .filter((i) => i !== null);
+}
+
+const impToPrebidBid =
+  (bidderRequestBody, bidResponseCurrency, referrer, targetingMap) => (bid) => {
+    try {
+      const bidRequest = findBidRequest(bidderRequestBody, bid);
+
+      if (!bidRequest) {
+        utils.logError(`${BIDDER_CODE}: Could not match bid to bid request`);
+
+        return null;
+      }
+      const categories = utils.deepAccess(bid, 'cat', []);
+      const mediaType = getMediaType(bid.adm);
+      let prebidBid = {
+        ad: bid.adm,
+        adId: bid.adid,
+        adserverTargeting: targetingMap[bid.impid],
+        adUnitCode: bidRequest.tagid,
+        bidderRequestId: bidderRequestBody.id,
+        bidId: bid.id,
+        cpm: bid.price,
+        creativeId: bid.crid || bid.id,
+        currency: bidResponseCurrency || getBidderConfig('defaultCurrency'),
+        height: bid.h,
+        lineItemId: utils.deepAccess(bid, 'ext.liid'),
+        mediaType,
+        meta: {
+          advertiserDomains: bid.adomain,
+          mediaType,
+          primaryCatId: categories[0],
+          secondaryCatIds: categories.slice(1),
+        },
+        netRevenue: true,
+        nurl: bid.nurl,
+        referrer: referrer,
+        requestId: bid.impid,
+        ttl: getBidderConfig('bidTTL'),
+        width: bid.w,
+      };
+
+      if (mediaType === VIDEO) {
+        prebidBid = {
+          ...prebidBid,
+          ...getVideoSpecificParams(bidRequest, bid),
+        };
+      }
+
+      return prebidBid;
+    } catch (error) {
+      utils.logError(`${BIDDER_CODE}: Error while building bid`, error);
+
+      return null;
+    }
+  };
+
+function getVideoSpecificParams(bidRequest, bid) {
+  return {
+    height: bid.h || bidRequest.video.h,
+    vastXml: bid.adm,
+    width: bid.w || bidRequest.video.w,
+  };
+}
+
+function buildTargetingMap(bids) {
+  const impIds = bids.map(({ impid }) => impid).filter(utils.uniques);
+  const values = impIds.reduce((result, id) => {
+    result[id] = {
+      lineItemIds: [],
+      dealIds: [],
+      adIds: [],
+    };
+
+    return result;
+  }, {});
+
+  bids.forEach((bid) => {
+    values[bid.impid].lineItemIds.push(bid.ext.liid);
+    values[bid.impid].dealIds.push(bid.dealid);
+    values[bid.impid].adIds.push(bid.adid);
+  });
+
+  const targetingMap = {};
+
+  for (const id of impIds) {
+    targetingMap[id] = {
+      hb_liid_adbookpsp: values[id].lineItemIds.join(TARGETING_VALUE_SEPARATOR),
+      hb_deal_adbookpsp: values[id].dealIds.join(TARGETING_VALUE_SEPARATOR),
+      hb_adid_c_adbookpsp: values[id].adIds.join(TARGETING_VALUE_SEPARATOR),
+    };
+  }
+
+  return targetingMap;
+}
+
+/**
+ * VALIDATION
+ */
+
+function hasRequiredParams(bidRequest) {
+  const value =
```

Comment:
This condition is only checking for nulls. This will return `true` if all 3 are `undefined`. This can be changed to the following, unless `0` needs to be accepted as a valid placementId or orgId:
```
utils.deepAccess(bidRequest, 'params.placementId') ||  utils.deepAccess(bidRequest, 'params.orgId') || getBidderConfig('orgId')
```

Reply:
It's actually checking for both `null` and `undefined`. Using `!== null` would check for nulls only, `!= null` returns `false` for both `null` and `undefined` since it ignores types.